[Substrate.io] Chain Specification (#27)
polkadot-developers/polkadot-docs: develop/parachain-devs/parachain-launch/chain-spec.md

---
title: Chain Specification

> description: Describes the role of the chain specification in a network, how to specify the chain specification to use when starting a node, and how to customize and distribute chain specifications.

Reviewer: The description is too long. It has 185 characters. It should be between 120 and 160

---

A _chain specification_ is the collection of information that describes a Polkadot SDK-based network. When starting a node, a chain specification is a crucial parameter, providing the genesis configurations, bootnodes, and other parameters relating to that particular network. It identifies the network that a blockchain node connects to, the other nodes that it initially communicates with, and the initial state that nodes must agree on to produce blocks.

> The chain specification is defined using the [`ChainSpec` struct](https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/struct.GenericChainSpec.html){target=_blank}.

Reviewer 2: nit: Maybe it would be better to describe a json that describes chain-spec?
Not sure if divin into rust structures makes sense.

But this depends on who is the target audience of this spec.
There is a description of fields here:
https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/index.html#intro-chain-specification

What reminds me that we need a json vaidation scheme for chain spec file (of course out of scope of this PR).

Reviewer: +1. I think it is better to refer to the Polkadot SDK docs instead of the Rust struct

> The chain specification is defined using the [`ChainSpec` struct](https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/struct.GenericChainSpec.html){target=_blank}.

Reviewer: ```suggestion
The chain specification is defined using the [`ChainSpec` struct](https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/struct.GenericChainSpec.html){target=\_blank}.
```

!!!note "This page has a companion reference document"
    To learn more about the particularities about chain specifications (and more specifically, the `chain-spec-builder` tool), [refer to the reference document.](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/chain_spec_genesis/index.html){target=_blank}

The [`ChainSpec` struct](https://paritytech.github.io/polkadot-sdk/master/parachain_template_node/chain_spec/type.ChainSpec.html){target=_blank} separates the information required for a chain into two parts:

> - A client specification that contains information used by the _outer node_ to communicate with network participants and send data to telemetry endpoints. Many of these chain specification settings can be overridden by command-line options when starting a node or can be changed after the blockchain has started

Reviewer 2: outer node -> node?

> - A client specification that contains information used by the _outer node_ to communicate with network participants and send data to telemetry endpoints. Many of these chain specification settings can be overridden by command-line options when starting a node or can be changed after the blockchain has started

Reviewer: ```suggestion
- A client specification that contains information used by the _node_ to communicate with network participants and send data to telemetry endpoints. Many of these chain specification settings can be overridden by command-line options when starting a node or can be changed after the blockchain has started
```

- The initial **genesis state** that all nodes in the network agree on. The genesis state must be established when the blockchain is first started and it cannot be changed thereafter without starting an entirely new blockchain

> ## Customizing Outer Node Settings

Reviewer: ```suggestion
## Customizing Node Settings
```

> For the outer node, the chain specification controls information such as:

Reviewer: ```suggestion
For the node, the chain specification controls information such as:
```

> For the outer node, the chain specification controls information such as:

Reviewer: Replace other occurrences of outer node

Note that you can customize outer node settings after genesis. However, nodes only add peers that use the same [`protocolId`](https://paritytech.github.io/polkadot-sdk/master/polkadot_parachain_lib/chain_spec/trait.ChainSpec.html#tymethod.protocol_id){target=_blank}.

## Customizing the Genesis Configuration

All nodes in the network must agree on the genesis state before they can agree on any subsequent blocks. The information configured in the genesis portion of a chain specification is used to create a genesis block. It takes effect when you start the first node and cannot be overridden with command-line options. However, you can configure some information in the genesis portion of a chain specification. For example, you can customize the genesis portion of the chain specification to include information such as:

- Initial account balances
- Accounts that are initially part of a governance council
- The account that controls the `sudo` key

> - Any other genesis state for a pallet

Reviewer 2: I would mention that this customization is feasible for development. For production raw format is recommended.

> Nodes also include the compiled WebAssembly for the runtime logic on the chain, so the initial runtime must also be supplied in the chain specification.

Reviewer 2: ```suggestion
Nodes also require the compiled WebAssembly for executing the runtime logic on the chain, so the initial runtime must also be supplied in the chain specification.
```

## Declaring storage items for a runtime

> In most cases, a runtime requires some storage items to be configured at genesis. This includes the initial state for pallets, for example, how much balance `Alice` or `Bob` exists, or the account that has sudo permissions.

Reviewer: ```suggestion
In most cases, a runtime requires some storage items to be configured at genesis. This includes the initial state for pallets, for example, how much balance `Alice` or `Bob` has, or the account that has sudo permissions.
```

> These storage values are configured in the genesis portion of the chain specification. You can either create a [patch](https://paritytech.github.io/polkadot-sdk/master/sc_chain_spec/index.html#chain-spec-formats){target=_blank}  file for [`chain-spec-builder`](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/chain_spec_genesis/index.html){target=_blank} to ingest, or if you're using `chain_spec.rs`, the pallet can be configured in the code.

Reviewer 2: I think link for chain-spec-builder is wrong.

But referring to the tutorial on genesis config (which is actually linked) may make sense.

<!-- TODO: Add a page for genesis configuration? -->

## Creating a Custom Chain Specification

There are two methods for creating a chain specification: 

1. Utilizing `polkadot-parachain` or compiled node binary
2. Utilizing `chain-spec-builder`

If you are creating a one-off network for development, testing, or demonstration purposes, you might want a fully customized chain specification.
To create a completely customized chain spec, you can export the default chain spec to JSON format, then edit the fields in the JSON file.

> For example, you can use the `build-spec` sub-command to export the chain specification to a JSON file:

Reviewer: I would remove the entire line and then just have the two subsections
```suggestion
```

### Using Node Binary

> !!!warning "Be aware of the Omninode"

Reviewer: Remove this. If chain spec builder is the preferred method, just put it first

>     As mentioned, there are two methods for generating a chain specification. The preferred method is [`chain-spec-builder`](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/chain_spec_genesis/index.html){target=_blank}, however the `build-spec` method can be utilized as well.

Reviewer 2: The link to chain-spec-builder is: https://paritytech.github.io/polkadot-sdk/master/staging_chain_spec_builder/index.html

Reviewer 2: And also why mention chain-spec-builder in `Using node binary` section? I'd remove it...

```bash
chain-spec-builder create -v -r <path-to-wasm-runtime-here>
```

## Raw chain specifications

With runtime upgrades, the blockchain's runtime can be upgraded with newer business logic. Chain specifications contain information structured in a way that can be understood by the node's runtime. For example, consider this excerpt of a common entry for a chain specification:

```json
"sudo": {
  "key": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
}
```

In the JSON file, this key and its associated value are human-readable text. However, this information can't be stored in this format in the underlying storage structures that Substrate uses. Before you can use the chain specification to initialize the genesis storage for a node, the human-readable keys must be transformed into actual storage keys that allow the values to be stored in the storage trie. This transformation is straight-forward, but it requires that the chain specification to be encoded in a format that node runtime can read.

> To enable a node with an upgraded runtime to synchronize with a chain from genesis, the human-readable chain specification is encoded in a **raw** format.

Reviewer 2: human-readable is often referred as `plain` (maybe worth using this phrasing too)

Reviewer: +1

The raw format enables you distribute chain specifications that all nodes can use to synchronize the chain even after runtime upgrades.

> Nodes support the `--raw` command-line option to produce the raw chain specifications.

Reviewer 2: Also `chain-spec-builder` supports that by:
 `create --raw` or dedicated `convert-to-raw` command.
